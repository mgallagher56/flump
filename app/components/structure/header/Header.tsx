import { type FC, type ReactElement, useEffect, useState } from 'react';

import { Flex } from '@chakra-ui/react';
import ColorModeSwitch from '~/components/ColorModeSwitch';
import FLPBox from '~/components/core/structure/FLPBox';
import NavMenu from '~/components/navigation/NavMenu';
import UserLogin from '~/components/navigation/UserLogin';

import HomeLogo from '../../navigation/HomeLogo';
import { loginStyles, menuStyles, navStyles } from './styles';

interface HeaderProps {
  showColorModeSwitch?: boolean;
}

const Header: FC<HeaderProps> = ({ showColorModeSwitch = true }): ReactElement => {
  const [isApp, setIsApp] = useState(false);
  useEffect(() => {
    setIsApp(!!window.location.pathname.includes('/app'));
  }, []);

  return (
    <FLPBox as="header">
      <Flex as="nav" className={navStyles}>
        <FLPBox className={menuStyles}>
          <HomeLogo />
          {isApp ? (
            <NavMenu
              routes={[
                { key: 'dashboard', route: '/app' },
                { key: 'accounts', route: '/app/accounts' }
              ]}
              showAppLink={false}
            />
          ) : (
            <NavMenu
              routes={[
                { key: 'home', route: '/' },
                { key: 'about', route: '/about' },
                { key: 'contact', route: '/contact' }
              ]}
            />
          )}
        </FLPBox>
        <FLPBox className={loginStyles}>
          <UserLogin />
          {showColorModeSwitch && <ColorModeSwitch />}
        </FLPBox>
      </Flex>
    </FLPBox>
  );
};

export default Header;
