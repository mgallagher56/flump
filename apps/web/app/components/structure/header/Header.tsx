import { type FC, type ReactElement } from 'react';

import { Flex } from '@chakra-ui/react';
import { useLoaderData } from 'react-router';
import ColorModeSwitch from '~/components/ColorModeSwitch';
import FLPBox from '~/components/core/structure/FLPBox';
import HomeLogo from '~/components/navigation/HomeLogo';
import NavMenu from '~/components/navigation/NavMenu';
import UserLogin from '~/components/navigation/UserLogin';
import type { loader } from '~/root';

import { loginStyles, menuStyles, navStyles } from './styles';

interface HeaderProps {
  showColorModeSwitch?: boolean;
}

const Header: FC<HeaderProps> = ({ showColorModeSwitch = true }): ReactElement => {
  const { user } = useLoaderData<typeof loader>();
  return (
    <FLPBox as="header">
      <Flex as="nav" className={navStyles}>
        <FLPBox className={menuStyles}>
          <HomeLogo />
          <NavMenu
            routes={[
              { key: 'home', route: '/' }
              // { key: 'about', route: '/about' },
              // { key: 'contact', route: '/contact' }
            ]}
          />
          {!!user && (
            <NavMenu
              routes={[
                { key: 'dashboard', route: '/app' },
                { key: 'accounts', route: '/app/accounts' }
              ]}
            />
          )}
        </FLPBox>
        <FLPBox className={loginStyles}>
          <UserLogin />
          {showColorModeSwitch && <ColorModeSwitch />}
        </FLPBox>
      </Flex>
    </FLPBox>
  );
};

export default Header;
