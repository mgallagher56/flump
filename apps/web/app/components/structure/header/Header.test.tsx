import { render } from '@testing-library/react';
import { vi } from 'vitest';

import mockUser from '__mocks__/user';

import Header from './Header';

const mocks = vi.hoisted(() => ({
  mockUseLoaderData: vi.fn()
}));

vi.mock('react-router', async () => {
  const actual: Record<string, unknown> = await vi.importActual('react-router');
  return {
    ...actual,
    useLoaderData: mocks.mockUseLoaderData
  };
});

vi.mock('app/components/ColorModeSwitch', () => ({
  default: () => <div>ColorModeSwitch</div>
}));

vi.mock('app/components/navigation/UserLogin', () => ({
  default: () => <div>UserLogin</div>
}));

vi.mock('app/components/navigation/NavMenu', () => ({
  default: ({ routes }) => routes.map(({ key }) => <div key={key}>{key}</div>)
}));

vi.mock('app/components/navigation/HomeLogo', () => ({
  default: () => <div>HomeLogo</div>
}));

describe('<Header />', () => {
  test('renders as expected by default', () => {
    mocks.mockUseLoaderData.mockReturnValue({ user: undefined });
    const { baseElement } = render(<Header />);
    expect(baseElement).toMatchSnapshot();
  });
  test('renders without ColorModeSwitch', () => {
    const { baseElement } = render(<Header />);
    expect(baseElement).toMatchSnapshot();
  });

  test('renders correctly when user logged in', () => {
    mocks.mockUseLoaderData.mockReturnValue({ user: mockUser });
    const { baseElement } = render(<Header />);
    expect(baseElement).toMatchSnapshot();
  });
});
