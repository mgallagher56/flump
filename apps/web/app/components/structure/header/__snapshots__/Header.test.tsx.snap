// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Header /> > renders as expected by default 1`] = `
<body>
  <div>
    <script
      nonce=""
    >
      ((e,r,s,u,d,m,l,h)=&gt;
      {let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;
      {let k=y==="class",S=k&&m?d.map(f=&gt;
      m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)
    </script>
    <header
      class="css-0"
    >
      <nav
        class="d_flex bd-b-w_2px bd-b-c_blue.500 border-bottom-style_solid ai_center jc_space-between w_100% py_4 px_2 css-k008qs"
      >
        <div
          class="d_flex gap_4 ai_center jc_start css-0"
        >
          <div>
            HomeLogo
          </div>
          <div>
            home
          </div>
        </div>
        <div
          class="d_flex gap_2 ai_center jc_end css-0"
        >
          <div>
            UserLogin
          </div>
          <button
            aria-label="Toggle color mode"
            class="chakra-button css-tf883e"
            type="button"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="4"
              />
              <path
                d="M12 2v2"
              />
              <path
                d="M12 20v2"
              />
              <path
                d="m4.93 4.93 1.41 1.41"
              />
              <path
                d="m17.66 17.66 1.41 1.41"
              />
              <path
                d="M2 12h2"
              />
              <path
                d="M20 12h2"
              />
              <path
                d="m6.34 17.66-1.41 1.41"
              />
              <path
                d="m19.07 4.93-1.41 1.41"
              />
            </svg>
          </button>
        </div>
      </nav>
    </header>
  </div>
</body>
`;

exports[`<Header /> > renders correctly when user logged in 1`] = `
<body>
  <div>
    <script
      nonce=""
    >
      ((e,r,s,u,d,m,l,h)=&gt;
      {let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;
      {let k=y==="class",S=k&&m?d.map(f=&gt;
      m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)
    </script>
    <header
      class="css-0"
    >
      <nav
        class="d_flex bd-b-w_2px bd-b-c_blue.500 border-bottom-style_solid ai_center jc_space-between w_100% py_4 px_2 css-k008qs"
      >
        <div
          class="d_flex gap_4 ai_center jc_start css-0"
        >
          <div>
            HomeLogo
          </div>
          <div>
            home
          </div>
          <div>
            dashboard
          </div>
          <div>
            accounts
          </div>
        </div>
        <div
          class="d_flex gap_2 ai_center jc_end css-0"
        >
          <div>
            UserLogin
          </div>
          <button
            aria-label="Toggle color mode"
            class="chakra-button css-tf883e"
            type="button"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="4"
              />
              <path
                d="M12 2v2"
              />
              <path
                d="M12 20v2"
              />
              <path
                d="m4.93 4.93 1.41 1.41"
              />
              <path
                d="m17.66 17.66 1.41 1.41"
              />
              <path
                d="M2 12h2"
              />
              <path
                d="M20 12h2"
              />
              <path
                d="m6.34 17.66-1.41 1.41"
              />
              <path
                d="m19.07 4.93-1.41 1.41"
              />
            </svg>
          </button>
        </div>
      </nav>
    </header>
  </div>
</body>
`;

exports[`<Header /> > renders without ColorModeSwitch 1`] = `
<body>
  <div>
    <script
      nonce=""
    >
      ((e,r,s,u,d,m,l,h)=&gt;
      {let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;
      {let k=y==="class",S=k&&m?d.map(f=&gt;
      m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)
    </script>
    <header
      class="css-0"
    >
      <nav
        class="d_flex bd-b-w_2px bd-b-c_blue.500 border-bottom-style_solid ai_center jc_space-between w_100% py_4 px_2 css-k008qs"
      >
        <div
          class="d_flex gap_4 ai_center jc_start css-0"
        >
          <div>
            HomeLogo
          </div>
          <div>
            home
          </div>
        </div>
        <div
          class="d_flex gap_2 ai_center jc_end css-0"
        >
          <div>
            UserLogin
          </div>
          <button
            aria-label="Toggle color mode"
            class="chakra-button css-tf883e"
            type="button"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="4"
              />
              <path
                d="M12 2v2"
              />
              <path
                d="M12 20v2"
              />
              <path
                d="m4.93 4.93 1.41 1.41"
              />
              <path
                d="m17.66 17.66 1.41 1.41"
              />
              <path
                d="M2 12h2"
              />
              <path
                d="M20 12h2"
              />
              <path
                d="m6.34 17.66-1.41 1.41"
              />
              <path
                d="m19.07 4.93-1.41 1.41"
              />
            </svg>
          </button>
        </div>
      </nav>
    </header>
  </div>
</body>
`;
